Rubberduck.RegexAssistant/IRegularExpression.cs

```diff
@@ -0,0 +1,327 @@
+                return string.Format("{0} {1}.", Atom.Description, Quantifier.HumanReadable());
+            }
+        
+        }
+
+        public Quantifier Quantifier
+        {
+            get
+            {
+                return quant;
+            }
+        }
+
+        public bool TryMatch(string text, out string remaining)
+        {
+            // try to match the atom a given number of times.. 
+            throw new NotImplementedException();
+        }
+
+        public override bool Equals(object obj)
+        {
+            if (obj is SingleAtomExpression)
+            {
+                SingleAtomExpression other = obj as SingleAtomExpression;
+                return other.Atom.Equals(Atom) && other.Quantifier.Equals(Quantifier);
+            }
+            return false;
+        }
+    }
+        
+    public static class RegularExpression
+    {
+
+        /// <summary>
+        /// We basically run a Chain of Responsibility here.At the outermost level, we need to check whether this is an AlternativesExpression.
+        /// If it isn't, we assume it's a ConcatenatedExpression and proceed to create one of these.
+        /// The next step is attempting to parse Atoms. Those are packed into a SingleAtomExpression with their respective Quantifier.
+        /// Note that Atoms can request a Parse of their subexpressions. Prominent example here would be Groups.
+        /// Also note that this here is responsible for separating atoms and Quantifiers. When we matched an Atom we need to try to match a Quantifier and pack them together. 
+        /// If there is no Quantifier following (either because the input is exhausted or there directly is the next atom) then we instead pair with `new Quantifier("")` 
+        /// </summary>
+        /// <param name="specifier"></param>
+        /// <returns></returns>
+        public static IRegularExpression Parse(string specifier)
+        {
+            IRegularExpression expression;
+            // ByRef requires us to hack around here, because TryParseAsAtom doesn't fail when it doesn't consume the specifier anymore
+            string specifierCopy = specifier;
+            if (TryParseAsAtom(ref specifierCopy, out expression) && specifierCopy.Length == 0)
+            {
+                return expression;
+            }
+            else
+            {
+                List<IRegularExpression> subexpressions = new List<IRegularExpression>();
+                while (specifier.Length != 0)
+                {
+                    expression = ParseIntoConcatenatedExpression(ref specifier);
+                    // ! actually an AlternativesExpression
+                    if (specifier.Length != 0 || subexpressions.Count != 0)
+                    {
+                        // flatten hierarchy
+                        var parsedSubexpressions = (expression as ConcatenatedExpression).Subexpressions;
+                        if (parsedSubexpressions.Count == 1)
+                        {
```

Comment:
Some of your conditions get nested pretty deeply--would it be possible to do early-exit (or early-break/continue) in some places to reduce nesting?